modules/host/contractmanager/sectoradd.go

```diff
@@ -0,0 +1,216 @@
+			// this sector.
+			//
+			// TODO: this search needs to consider storage folders that are in
+			// the process of being removed, and should avoid any storage
+			// folders which are undergoing a remove operation which is
+			// incomplete. This would look like a pre-process which compares
+			// the list of committed storage folders to the list of uncommitted
+			// removals, before passing the result to emptiestStorageFolder.
+			sfs := wal.storageFolders()
+			sf, _ := emptiestStorageFolder(sfs)
+			if sf == nil {
+				// None of the storage folders have enough room to house the
+				// sector.
+				return errInsufficientStorageForSector
+			}
+
+			// Find a location for the sector within the file using the Usage
+			// field.
+			sectorIndex := randFreeSector(sf.Usage)
+
+			// Set the sectorAdd fields and update the sectorLocations map to
+			// reflect this new sector.
+			sa.Count = 1
+			sa.Folder = sf.Index
+			sa.Index = sectorIndex
+			wal.cm.sectorLocations[id] = sectorLocation{
+				index:         sectorIndex,
+				storageFolder: sf.Index,
+				count:         1,
+			}
+		}
+
+		// Add a change to the WAL to commit this sector to the provided index.
+		err := wal.appendChange(stateChange{
+			AddedSectors: []sectorAdd{sa},
+		})
+		if err != nil {
+			return build.ExtendErr("failed to add a state change", err)
+		}
+
+		// Grab the synchronization channel so that we know when the sector add
+		// has completed.
+		syncChan = wal.syncChan
+		return nil
+	}()
+	if err != nil {
+		return build.ExtendErr("verification failed:", err)
+	}
+
+	// Only return after the commitment has succeeded fully.
+	<-syncChan
+	return nil
+}
+
+// commitAddSector will commit a sector that has been added to the WAL. The
+// commit should be idempotent, meaning if this function is run multiple times
+// on the same sector, there should be no issues.
+func (wal *writeAheadLog) commitAddSector(sa sectorAdd) {
+	sf := wal.cm.storageFolders[sa.Folder]
+	lookupTableSize := len(sf.Usage) * 64 * 16
+
+	// Write the sector metadata to disk.
+	writeData := make([]byte, 16)
+	copy(writeData, sa.ID[:])
+	binary.LittleEndian.PutUint16(writeData[12:14], sa.Count)
+	binary.LittleEndian.PutUint16(writeData[14:], sa.Folder)
+	_, err := sf.file.Seek(16*int64(sa.Index), 0)
+	if err != nil {
+		wal.cm.log.Println("ERROR: unable to seek to sector metadata when adding sector")
+		sf.FailedWrites += 1
+		return
+	}
+	_, err = sf.file.Write(writeData)
+	if err != nil {
+		sf.FailedWrites += 1
+		return
+	}
+
+	// Write the sector to disk. The write only needs to happen if the count is
+	// equal to 1, otherwise the data can be assumed to already be there.
+	if sa.Count == 1 {
+		_, err = sf.file.Seek(int64(modules.SectorSize)*int64(sa.Index)+int64(lookupTableSize), 0)
+		if err != nil {
+			wal.cm.log.Println("ERROR: unable to seek to sector data when adding sector")
+			sf.FailedWrites += 1
+			return
+		}
+		_, err = sf.file.Write(sa.Data)
```

Comment:
`os.File` supports `WriteAt`, which uses a different syscall (`pwrite`). It's only good for one-off writes, though; IIRC, `WriteAt` does not affect the write position of `Write`.